packages/@lwc/template-compiler/src/__tests__/fixtures/directive-inner-html/usage-if-for-each/expected.js

```diff
@@ -0,0 +1,71 @@
+import _aB from "a/b";
+import { registerTemplate, sanitizeHtmlContent } from "lwc";
+function tmpl($api, $cmp, $slotset, $ctx) {
+  const {
+    h: api_element,
+    k: api_key,
+    i: api_iterator,
+    f: api_flatten,
+    c: api_custom_element,
+  } = $api;
+  return [
+    api_custom_element(
+      "a-b",
+      _aB,
+      {
+        classMap: {
+          s2: true,
+        },
+        key: 0,
+      },
+      api_flatten([
+        $cmp.isTrue
+          ? api_element(
+              "div",
+              {
+                props: {
+                  innerHTML:
+                    $ctx._rawHtml$0 !== ($ctx._rawHtml$0 = $cmp.ifRawHtml)
+                      ? ($ctx._sanitizedHtml$0 = sanitizeHtmlContent(
+                          $cmp.ifRawHtml
+                        ))
+                      : $ctx._sanitizedHtml$0,
+                },
+                context: {
+                  lwc: {
+                    dom: "manual",
+                  },
+                },
+                key: 1,
+              },
+              []
+            )
+          : null,
+        api_iterator($cmp.items, function (item) {
+          return api_element(
+            "div",
+            {
+              props: {
+                innerHTML:
+                  $ctx._rawHtml$1 !== ($ctx._rawHtml$1 = item.forRawHtml)
+                    ? ($ctx._sanitizedHtml$1 = sanitizeHtmlContent(
+                        item.forRawHtml
+                      ))
+                    : $ctx._sanitizedHtml$1,
```

Comment:
Note: In the foreach case, when the innerHTML depends on the for item, the optimized expr won't help much.

Reply:
Can you think of a good solution to this? I can't think of anything, except maybe modifying the `_rawHtml$1`/`_sanitizedHtml$1` key to include the `i` from the iteration. (E.g. `_rawHtml$1_0`, `_rawHtml$1_1`, `_rawHtml$1_2`, etc.) But then it would only help if the innerHTML depends on the item; it'd be wasteful otherwise.

Reply:
Another possibility:

```js
function tmpl($api, $cmp, $slotset, $ctx) {
  const cache = {}
  const sanitize = (input) => {
    let output = cache[input]
    if (!output) {
      cache[input] = output = sanitizeHtmlContent(input)
    }
    return output
  }
  /* ... */
  api_iterator($cmp.items, function (item) {
    /* ... */
    innerHTML: sanitize(item.forRawHtml)
  }
}
```

This way, it would be cached no matter where it's called inside the `tmpl` function. Also the `cache` should not leak – it should only be alive while `tmpl` is being called.

Reply:
IMHO, I would not recommend using the `lwc:inner-html` directive within an iteration, I'm struggling to think of a good use case for it, at least one using a large text.

About your proposed solutions:

The first one, the values won't be collected from `$ctx`.

The second solution should not work because when the template is evaluated between renders, the cache will always be empty, therefore `sanitizeHtmlContent` will be called multiple times.

Reply:
I think you're right, we can just leave it as-is and let `lwc:inner-html` be a bit unoptimized when used in a loop. It's probably an edge-case anyway.